subprojects/docs/src/docs/userguide/tasks.xml

```diff
@@ -290,52 +290,673 @@
+                    <sample id="incrementalAdHocTask"
+                            dir="userguide/tasks/incrementalBuild/customTaskClass"
+                            title="Ad-hoc task">
+                        <sourcefile file="build.gradle" snippet="ad-hoc-task"/>
+                        <output args="processTemplatesAdHoc" ignoreExtraLines="true" ignoreLineOrder="true"/>
+                        <output args="processTemplatesAdHoc" ignoreExtraLines="true" ignoreLineOrder="true"
+                                outputFile="incrementalAdHocTaskUpToDate.out"/>
+                    </sample>
+                    <para>As before, there’s much to talk about. To begin with, you should really write
+                        a custom task class for this as it’s a non-trivial implementation that has
+                        several configuration options. In this case, there are no task properties to
+                        store the root source folder, the location of the output directory or any of
+                        the other settings. That’s deliberate to highlight the fact that the runtime
+                        API doesn’t require the task to have any state. In terms of incremental build,
+                        the above ad-hoc task will behave the same as the custom task class.
+                    </para>
+                    <para>All the input and output definitions are done through the methods on
+                        <literal>inputs</literal> and <literal>outputs</literal>, such as
+                        <literal>property()</literal>, <literal>files()</literal>, and
+                        <literal>dir()</literal>. Gradle performs up-to-date checks on the argument
+                        values to determine whether the task needs to run again or not. Each method
+                        corresponds to one of the incremental build annotations, for example
+                        <literal>inputs.property()</literal> maps to <literal>@Input</literal> and
+                        <literal>outputs.dir()</literal> maps to <literal>@OutputDirectory</literal>.
+                        The only difference is that the <literal>file()</literal>, <literal>files()</literal>,
+                        <literal>dir()</literal> and <literal>dirs()</literal> methods don’t validate
+                        the type of file object at the given path (file or directory), unlike the annotations.
+                    </para>
+                    <para>One notable hole in the runtime API currently is the lack of support for nested
+                        values. That’s why the example uses two <literal>property()</literal> declarations
+                        for the template data. 
+                    </para>
+                </section>
+
+                <section id="sec:runtime_api_for_custom_tasks">
+                    <title>Using it for custom task types</title>
```

Comment:
We usually say "custom task" if you create your own task type implementation - pretty much any class that extends `DefaultTask`. On contrary we'd just say "task type" if it is provided by the Gradle core API. Ad-hoc tasks are tasks you define in your build script but do not explicitly declare a type. We call it add-hoc tasks because it's a one-off implementation, where as a custom task implementation is reusable functionality. I think we should clarify these definitions in the chapter about tasks (not necessarily as part of this pull request).